src/orders/service/order.service.ts

```diff
@@ -0,0 +1,265 @@
+import { OrderDetail } from '../domain/order-detail.entity';
+import { OrderRepository } from '../repository/order.repository';
+import { ProductService } from 'src/products/service/product.service';
+import { Stock } from 'src/products/domain/stock.entity';
+import { StockService } from 'src/products/service/stock.service';
+
+@Injectable()
+export class OrderService {
+  constructor(
+    private readonly orderRepository: OrderRepository,
+    private readonly customerService: CustomerService,
+    private readonly stockService: StockService,
+    private readonly productService: ProductService,
+    private readonly dataSource: DataSource,
+  ) {}
+
+  async getById(orderId: string): Promise<Order> {
+    const order = await this.orderRepository.findById(orderId);
+    if (!order) {
+      throw new NotFoundException('No order found by given ID');
+    }
+
+    return order;
+  }
+
+  async getOrdersByCustomerId(customerId: string): Promise<Order[]> {
+    const orders =
+      await this.orderRepository.findOrdersByCustomerId(customerId);
+    if (orders.length === 0) {
+      throw new NotFoundException(
+        'No orders found containing the desired product',
+      );
+    }
+
+    return orders;
+  }
+
+  async getOrdersByOrderedProductId(productId: string): Promise<Order[]> {
+    const orders =
+      await this.orderRepository.findOrdersByOrderedProductId(productId);
+    if (orders.length === 0) {
+      throw new NotFoundException(
+        'No orders found containing the desired product',
+      );
+    }
+
+    return orders;
+  }
+
+  async placeOrder(
+    newOrder: Omit<Order, 'id' | 'orderDetails' | 'customer' | 'createdAt'>,
+    customerId: string,
+    desiredOrderItems: DesiredOrderItem[],
+  ): Promise<Order> {
+    const queryRunner = this.dataSource.createQueryRunner();
+    await queryRunner.connect();
+    await queryRunner.startTransaction();
+
+    try {
+      const customer = await this.customerService.getById(customerId);
+      const orderedItemsStocks =
+        await this.getOrderedItemsStocks(desiredOrderItems);
+
+      this.validateStockAvailability(orderedItemsStocks, desiredOrderItems);
+
+      const order = this.createOrderEntity(newOrder, customer);
+      const orderDetails = this.createOrderDetails(
+        order,
+        orderedItemsStocks,
+        desiredOrderItems,
+      );
+      order.orderDetails = orderDetails;
+      await queryRunner.manager.save(order);
+      await this.updateStocks(orderedItemsStocks, desiredOrderItems);
+
+      await queryRunner.commitTransaction();
+      return order;
+    } catch (error) {
+      await queryRunner.rollbackTransaction();
+      throw error;
+    } finally {
+      await queryRunner.release();
+    }
+  }
+
+  async updateOrder(
+    orderId: string,
+    updatedOrderData: Partial<Order>,
+    updatedOrderItems: DesiredOrderItem[],
+  ): Promise<Order> {
+    const queryRunner = this.dataSource.createQueryRunner();
```

Comment:
**suggestion(non-blocking):** note that using migration is getting quite verbose in most of the cases. For the future is good to share somehow this logic in the future (there are several ways, including using decorators (either custom or from a library: https://www.npmjs.com/package/typeorm-transactional))

But currently it is out-of-scope for this training, it is very good that you worked and gotten familiar with it.